staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscoveryList is a resource containing a list of APIGroupDiscoveries.
+// The shape is chosen to be very similar to "normal" api endpoints and work well with our type system
+// This is what is returned from the /discovery/v2 endpoint.
+type APIGroupDiscoveryList struct {
+	TypeMeta `json:",inline"`
+	// ResourceVersion will not be set, because this does not have a replayable ordering among multiple apiservers.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Groups is the list of groups for discovery.
+	// +listType=map
+	// +listMapKey=name
+	Groups []APIGroupDiscovery `json:"groups" protobuf:"bytes,2,rep,name=groups"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscovery holds information about which resources are being served for all version of the API Group.
+type APIGroupDiscovery struct {
+	TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// The only field completed will be Name.  For instance, resourceVersion will be empty.
```

Comment:
No :-/

Reply:
Given HA apiservers, there is no agreed upon ordering for resourceversions of these.  I think asking for persistence would be a significant change in how this is built.  I'm open to it, but I wouldn't require it.

Reply:
We could commit this to etcd (and then serve that version). I think we should do that at some point, but I don't think it needs to block this. It should be backwards-compatible to start setting e.g. resourceVersion and supporting WATCH on this in the future if we want to.

Reply:
Allowing watchable discovery would be pretty amazing - are you committed to a particular implementation @deads2k ?

Reply:
I'm a bit concerned watchable discovery didn't come up in the KEP before we started implementation.  That is a very common client request.  I don't really care whether "we can't implement it today", but I have a serious concern about implementing this API without a way to get to it.  I'd like to see an update to the KEP that calls out it should be possible and that an API design for that is a requirement to get to beta.

Reply:
So taking a second to respond directly (in the "this is the ideal solution"), if we:

1. Defined `/apis/v1/discovery.k8s.io/apiresources` and returned a list of `APIGroupDiscovery`
2. Clients called LIST and merged the results into their discovery object on their side
3. [completely arbitrary solution for etag alternative, 5 seconds of design thought] Created a singleton resource in etcd that was updated everytime we calculated a new hash (hash stored on the object as a field) and returned the resource version of that object from the apiserver when you make a LIST call, and used resource version semantics to handle rejection of the request (every server has a linear version) instead of etag

we would have a functionally equivalent solution that uses only Kubernetes concepts, can be a generated client, objects normal access rules for resources, and could have an impl that later on modifies the object impl and supports watch (by transitioning the calculation of the object history to a partial object, or simply cheating within watch semantics like we do on watch cache and jumping straight to newest api endpoints).  Which addresses your 3 key design criteria.   It also means that we would get API docs on APIGroupDiscovery (via explain) and if we have to evolve that endpoint in the future we can simply use all of our existing processes.

That is more complicated than the current proposal, but it is also more consistent, on a path that leads to consistent watch, and likely to reuse 90% of our current resource implementation.  It comes with some disadvantages - I could imagine that discovery as a resource causes some problems that discovery not as a resource has (similar to namespace), but it is less arbitary than the new halfway point.  Did we consider this approach (possibly not since it seems like no one represented potential watchability as long term desirable)?

Note here I'm very used to complex clients (user interfaces) complaining how inadequate the current discovery behavior is for dealing with dynamically added types, and so it may be that those voices simply never actually registered the request to Kube issues to sig api-machinery, in which case we're saying it's not a requirement, and that those users can just come along later and do the work with a different discovery impl.  Which I'm ok with in general, since you can't design for things people don't ask for.

Reply:
I should note that an argument of "we should NEVER do watch, because with this API polling with etag will be so easy and cheap no one will ever care" is persuasive - if that's an implicit argument you are making, i'm open to it if we can't find strong use cases for watch above and beyond poll.